File: java/143.reorder-list.java
/*
 * @lc app=leetcode id=143 lang=java
 *
 * [143] Reorder List
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        if (head == null || head.next == null) return;
        ListNode prev = null, slow = head, fast = head;
        while (fast != null && fast.next != null) {
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }
        prev.next = null;
        ListNode revHead = reverseList(slow);
        ListNode be4Head = new ListNode(0), ptr = be4Head;

        boolean flag = true;
        while (head != null && revHead != null) {
            if (flag) {
                ptr.next = head;
                head = head.next;
            } else {
                ptr.next = revHead;
                revHead = revHead.next;
            }
            ptr = ptr.next;
            flag = !flag;
        }
        if (head == null) ptr.next = revHead;
        if (revHead == null) ptr.next = head;
    }

    private ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) return head;
        ListNode prev = null, curr = head, next = head.next;
        while (curr != null) {
            curr.next = prev;
            prev = curr;
            curr = next;
            if (next != null) next = next.next;
        }
        return prev;
    }

    private void tester(ListNode head) {
        int itr = 0;
        while (itr++ < 10 && head != null) {
            System.out.print(head.val + "->");
            head = head.next;
        }
        System.out.println();
    }
}

File: go/143.reorder-list.go
/*
 * @lc app=leetcode id=143 lang=golang
 *
 * [143] Reorder List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reorderList(head *ListNode)  {
	if head == nil || head.Next == nil {
		return
	}
	ptr := head
	list := make([]*ListNode, 0)
    for ptr != nil {
		list = append(list, ptr)
		ptr = ptr.Next
	}
	for i := 0; i < len(list)/2; i++ {
		list[i].Next = list[len(list)-1-i]
		list[len(list)-1-i].Next = list[i+1]
	}
	list[len(list)/2].Next = nil
}
// @lc code=end

